### PR TITLE
Writeup on C# Language changes in async2 experiment.

### PR DESCRIPTION
**Pull request by VSadov**

### CODE REVIEW

File: docs/design/features/runtime-handled-tasks.md [hunk 1]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+    if (!awaiter.IsCompleted)
+    {
+        RuntimeHelpers.UnsafeAwaitAwaiterFromRuntimeAsync<YieldAwaitable.YieldAwaiter>(awaiter);
+    }
+    awaiter.GetResult();
+    return 3;
+}
+```
+
+Operations that are not affected by `async2`, such as generic substitution or Overriding/Hiding/Implementing end up naturally working with async2 methods. Note that these areas are some of the most complex parts of the compiler. It was very benefitial to not having to specialcase async2 in these areas.
```
Comment (danmoseley) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
Operations that are not affected by `async2`, such as generic substitution or Overriding/Hiding/Implementing end up naturally working with async2 methods. Note that these areas are some of the most complex parts of the compiler. It was very beneficial to not having to specialcase async2 in these areas.
```

File: docs/design/features/runtime-handled-tasks.md [hunk 2]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+#### Unresolved concerns.
+None of the following appears to be unresolvable or blocking, we just did not get to these due to time constraints and the scoping of the experiment.
+
+1. There is an existing concept for `async void` methods. Does that model need to exist in the new system?
+1. It looks like there might be a need for async2 flavors of delegates. 
+To the runtime these would look like delegates with async2 `Invoke` method and may "just work". They would need some kind of sintax in C#.
+1. It looks like there might be a need for async2 flavors of method pointers - to represent async2 methods. 
+In the runtime this is representable via the modopt on the return type and, again, might "just work", so this is mostly a matter of picking C# syntax/semantics.
+1. Is there any impact on async enumerable/iterator?  
+While async2 is interoperable with regular async and these interfaces could be used or even implemented by the means of async2 methods, there could be opprtunities to employ async2 implementaitons in the lowering. It is not clear if that could be beneficial without defining/deriving async2 versions of `IAsyncEnumerable`/`IAsyncEnumerator`/`IAsyncDisposable`.
```
Comment (danmoseley) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
While async2 is interoperable with regular async and these interfaces could be used or even implemented by the means of async2 methods, there could be opprtunities to employ async2 implementations in the lowering. It is not clear if that could be beneficial without defining/deriving async2 versions of `IAsyncEnumerable`/`IAsyncEnumerator`/`IAsyncDisposable`.
```

File: docs/design/features/runtime-handled-tasks.md [hunk 3]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+
+#### ==== Syntax v1 (abandoned)
+
+Initialy we assumed a different syntax for async2 methods. It ended up not working well beyond simple scenarios. We will still describe that here for future reference as something that appear to be a syntactical dead-end.
```
Comment (danmoseley) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
Initially we assumed a different syntax for async2 methods. It ended up not working well beyond simple scenarios. We will still describe that here for future reference as something that appear to be a syntactical dead-end.
```

File: docs/design/features/runtime-handled-tasks.md [hunk 4]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+async2 int M1(int arg) {}
+
+// meaning in IL 
+int int32 modopt([System.Runtime]System.Threading.Tasks.Task`1) M1(int arg)
+```
 
-1. What about code which today looks like `public async Task<Task<Task<int>>> Method()` What would the new syntax be?
-2. It is desirable to be able to implement methods that return `ValueTask` as well as just `Task` with an async2 method. How do we distinguish?
-3. There is an existing concept for `async void` methods. Does that model need to exist in the new system.
+While intially favored for being close to the IL representation, it was eventually abandoned due to major problems and inconsistencies.
```
Comment (danmoseley) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
While initially favored for being close to the IL representation, it was eventually abandoned due to major problems and inconsistencies.
```

File: docs/design/features/runtime-handled-tasks.md [hunk 5]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+
+One thing to observe is that unlike regular `async`, which is a source-only concept, the async2 survives serialization/deserialization via metadata. 
+Effectively we treat the `int int32 modopt([System.Runtime]System.Threading.Tasks.Task'1)` as just a special encoding of `Task<int>` return type with additional property of making the method `async2`.
+
+#### Unresolved concerns.
+None of the following appears to be unresolvable or blocking, we just did not get to these due to time constraints and the scoping of the experiment.
+
+1. There is an existing concept for `async void` methods. Does that model need to exist in the new system?
+1. It looks like there might be a need for async2 flavors of delegates. 
+To the runtime these would look like delegates with async2 `Invoke` method and may "just work". They would need some kind of sintax in C#.
```
Comment (jkotas) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
To the runtime these would look like delegates with async2 `Invoke` method and may "just work". They would need some kind of syntax in C#.
```

File: docs/design/features/runtime-handled-tasks.md (hunk 2 above)
Comment (jkotas) [inline docs/design/features/runtime-handled-tasks.md]:
```suggestion
While async2 is interoperable with regular async and these interfaces could be used or even implemented by the means of async2 methods, there could be opportunities to employ async2 implementaitons in the lowering. It is not clear if that could be beneficial without defining/deriving async2 versions of `IAsyncEnumerable`/`IAsyncEnumerator`/`IAsyncDisposable`.
```

File: docs/design/features/runtime-handled-tasks.md [hunk 6]
```diff
@@ -452,13 +455,104 @@ async2 ReturnType Thunk(ParameterType param1, ParameterType2 param2, ...)
...
+
+Initialy we assumed a different syntax for async2 methods. It ended up not working well beyond simple scenarios. We will still describe that here for future reference as something that appear to be a syntactical dead-end.
+
+The initial version of the syntax looked like:
+
+```cs
+async2 int M1(int arg) {}
+
+// meaning in IL 
+int int32 modopt([System.Runtime]System.Threading.Tasks.Task`1) M1(int arg)
```
Comment (jkotas) [inline docs/design/features/runtime-handled-tasks.md]: I am not following this example. What does `int int32` return type mean?

Comment (VSadov) [reply docs/design/features/runtime-handled-tasks.md]: Oh, its a typo. Should be just int32 with a modopt. I had the same typo in a few places and fixed. Somehow missed this one.

**Review by davidwrighton — APPROVED** [review]

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**